api/xds_protocol.rst

```diff
@@ -385,23 +385,33 @@ from the client's perspective.
 Knowing When a Requested Resource Does Not Exist
 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
-In the SotW protocol variants, responses for :ref:`Listener <envoy_api_msg_Listener>` and
-:ref:`Cluster <envoy_api_msg_Cluster>` resource types must include all resources requested by the
-client. Therefore, if a client requests a resource that does not exist, it can immediately
-tell this from the response.
-
-However, for other resource types, because each resource can be sent in its own response, there is
-no way to know from the next response whether the newly requested resource exists, because the next
+The SotW protocol variants do not provide any explicit mechanism to determine when a requested
+resource does not exist.
+
+Responses for :ref:`Listener <envoy_api_msg_Listener>` and :ref:`Cluster <envoy_api_msg_Cluster>`
+resource types must include all resources requested by the client. However, there is still a
+race condition here: if the client initially sends a request for resource A, then sends a request
+for resources A and B, and then sees a response containing only resource A, the client cannot
+conclude that resource B does not exist, because the response may have been sent on the
+basis of the first request, before the server saw the second request.
```

Comment:
I think we should just fix this with request nonces. I'm uncomfortable with documenting broken behavior as intended. This was never an issue in Envoy, since CDS/LDS never have a specified resource in the request.

Reply:
How exactly would "request nonces" work?  Is the idea that the client would send a nonce in the initial request and the server would somehow echo it back?  I'd like to see a specific proposal for the semantics to better understand how invasive this should be.

Another option would be to say that when LDS or CDS is in non-wildcard mode, it does not need to include all resoures in every response -- i.e., it would behave like RDS and EDS and provide each resource in its own response.  But that also seems like a fairly invasive change to the existing protocol.

Note that even before this PR, we said that the client would need to use a timeout anyway.  So all we're really saying here is that if the client is using non-wildcard mode for LDS or CDS, it can't short-circuit the timeout when it sees the next response.  This seems like the best way to minimize the changes needed to existing management servers to allow gRPC clients to interact with them.

Reply:
> How exactly would "request nonces" work? Is the idea that the client would send a nonce in the initial request and the server would somehow echo it back? I'd like to see a specific proposal for the semantics to better understand how invasive this should be.
> 

"request nonces" would need more fields in `DiscoveryRequest` and `DiscoveryResponse` right? It wouldn't be a good idea to overload the existing `nonce` and `response_nonce` fields for this purpose. If more fields are going to be added I suggest also adding `error_detail` in `DiscoveryResponse` so the management server could respond with errors like `NOT_FOUND` or `INVALID_ARGUMENT` to address some of these problems.

Reply:
Yes, additional fields. 

> Is the idea that the client would send a nonce in the initial request and the server would somehow echo it back?

Yes. This is how response nonces work. It's pretty simple and fixes the race condition.

My main objection here is how we're characterizing new behavior that we're designing as inherently racy. If instead it's the case that it's "eventual consistent" and we need to wait for a timeout, that's all good.

Reply:
Yeah, my intent here was just to clarify the wording, not really to make a substantive change.  The previous wording implied that for LDS and CDS, the client could know that a resource does not exist by virtue of it being missing in a response, but this race condition makes it impossible.  So that just means that we need to use the timeout for LDS and CDS, just like we do for any other resource type -- and just like the previous wording recommended for LDS and CDS anyway, to catch the case where the management server doesn't send a response in a timely manner.

It is still okay to know that a resource doesn't exist based on the LDS or CDS response in wildcard mode, which is what Envoy uses.  So I don't think this really changes much.

I've attempted to clarify the wording here.